source/fem-dem/void_fraction.cc

```diff
@@ -0,0 +1,1141 @@
+    dsp,
+    locally_owned_dofs,
+    this->triangulation->get_communicator(),
+    locally_relevant_dofs);
+
+  system_matrix_void_fraction.reinit(locally_owned_dofs,
+                                     locally_owned_dofs,
+                                     dsp,
+                                     this->triangulation->get_communicator());
+
+
+  system_rhs_void_fraction.reinit(locally_owned_dofs,
+                                  this->triangulation->get_communicator());
+
+
+  // Vertices to cell mapping
+  LetheGridTools::vertices_cell_mapping(this->dof_handler, vertices_to_cell);
+}
+
+
+template <int dim>
+void
+VoidFractionBase<dim>::setup_constraints(
+  const BoundaryConditions::NSBoundaryConditions<dim> &boundary_conditions)
+{
+  has_periodic_boundaries = false;
+  // Define constraints for periodic boundary conditions
+  void_fraction_constraints.clear();
+  void_fraction_constraints.reinit(locally_relevant_dofs);
+  DoFTools::make_hanging_node_constraints(dof_handler,
+                                          void_fraction_constraints);
+
+
+  for (auto const &[id, type] : boundary_conditions.type)
+    {
+      if (type == BoundaryConditions::BoundaryType::periodic)
+        {
+          periodic_direction = boundary_conditions.periodic_direction.at(id);
+          DoFTools::make_periodicity_constraints(
+            this->dof_handler,
+            id,
+            boundary_conditions.periodic_neighbor_id.at(id),
+            periodic_direction,
+            this->void_fraction_constraints);
+
+          has_periodic_boundaries = true;
+
+          // Get periodic offset if void fraction method is qcm or spm
+          if (this->void_fraction_parameters->mode ==
+                Parameters::VoidFractionMode::qcm ||
+              this->void_fraction_parameters->mode ==
+                Parameters::VoidFractionMode::spm)
+            {
+              periodic_offset = get_periodic_offset_distance(id);
+            }
+        }
+    }
+  void_fraction_constraints.close();
+
+
+  // Reinit system matrix
+
+  DynamicSparsityPattern dsp(locally_relevant_dofs);
+  DoFTools::make_sparsity_pattern(dof_handler,
+                                  dsp,
+                                  void_fraction_constraints,
+                                  false);
+
+  SparsityTools::distribute_sparsity_pattern(
+    dsp,
+    locally_owned_dofs,
+    this->triangulation->get_communicator(),
+    locally_relevant_dofs);
+
+  system_matrix_void_fraction.reinit(locally_owned_dofs,
+                                     locally_owned_dofs,
+                                     dsp,
+                                     this->triangulation->get_communicator());
+
+  if (has_periodic_boundaries)
+    LetheGridTools::vertices_cell_mapping_with_periodic_boundaries(
+      this->dof_handler, this->vertices_to_periodic_cell);
+}
+
+
+template <int dim>
+void
+VoidFractionBase<dim>::calculate_void_fraction(const double time)
```

Comment:
Could time be passed by reference?

Reply:
yes but for a double that does not make a difference. The length of a pointer or a double is about the same